File: packages/sdk-python/src/zroky/monitor.py
"""ZROKY Monitor — high-level async event tracker with batching and fail-open."""

from __future__ import annotations

import atexit
import logging
import threading
import time
from typing import Any

from zroky.client import ZrokyClient

logger = logging.getLogger("zroky")

_DEFAULT_BATCH_SIZE = 50
_DEFAULT_FLUSH_INTERVAL = 5.0
_DEFAULT_MAX_QUEUE = 10_000
_CIRCUIT_BREAKER_THRESHOLD = 5
_CIRCUIT_BREAKER_COOLDOWN = 60.0


class ZROKYMonitor:
    """High-level monitor that batches events and sends them asynchronously.

    Fail-open by default — if ZROKY is unreachable, client AI keeps working.
    Events are queued locally and sent in background.
    """

    def __init__(
        self,
        api_key: str,
        agent_id: str,
        *,
        fail_open: bool = True,
        batch_size: int = _DEFAULT_BATCH_SIZE,
        flush_interval: float = _DEFAULT_FLUSH_INTERVAL,
        max_queue_size: int = _DEFAULT_MAX_QUEUE,
        base_url: str = "https://api.zroky.ai",
        ingest_url: str = "https://ingest.zroky.ai",
    ) -> None:
        if not api_key:
            raise ValueError("api_key is required")
        if not agent_id:
            raise ValueError("agent_id is required")

        self._agent_id = agent_id
        self._fail_open = fail_open
        self._batch_size = batch_size
        self._flush_interval = flush_interval
        self._max_queue_size = max_queue_size

        self._client = ZrokyClient(api_key=api_key, base_url=base_url, ingest_url=ingest_url)
        self._queue: list[dict[str, Any]] = []
        self._lock = threading.Lock()
        self._closed = False

        # Circuit breaker state
        self._consecutive_failures = 0
        self._circuit_open_until = 0.0

        # Background flush thread
        self._flush_thread = threading.Thread(target=self._flush_loop, daemon=True)
        self._flush_thread.start()

        atexit.register(self.close)

    def track(
        self,
        prompt: str,
        response: str,
        *,
        model: str = "",
        session_id: str = "",
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Track an AI interaction event. Returns immediately (< 1ms)."""
        if self._closed:
            raise RuntimeError("Monitor is closed")

        event = {
            "agent_id": self._agent_id,
            "prompt": prompt,
            "response": response,
            "model": model,
            "session_id": session_id,
            "metadata": metadata or {},
            "timestamp": time.time(),
        }

        with self._lock:
            if len(self._queue) >= self._max_queue_size:
                if self._fail_open:
                    logger.warning("ZROKY queue full (%d), dropping event", self._max_queue_size)
                    return {"status": "dropped", "reason": "queue_full"}
                raise RuntimeError(f"ZROKY event queue full ({self._max_queue_size})")
            self._queue.append(event)

        return {"status": "queued"}

    def flush(self) -> int:
        """Flush all queued events immediately. Returns count of events sent."""
        with self._lock:
            if not self._queue:
                return 0
            batch = self._queue[: self._batch_size]
            self._queue = self._queue[self._batch_size :]

        return self._send_batch(batch)

    def close(self) -> None:
        """Flush remaining events and close the monitor."""
        if self._closed:
            return
        self._closed = True

        # Flush all remaining
        while True:
            with self._lock:
                if not self._queue:
                    break
                batch = self._queue[: self._batch_size]
                self._queue = self._queue[self._batch_size :]
            self._send_batch(batch)

        self._client.close()

    # -- Internal ------------------------------------------------------------

    def _flush_loop(self) -> None:
        """Background thread that periodically flushes the queue."""
        while not self._closed:
            time.sleep(self._flush_interval)
            try:
                while True:
                    with self._lock:
                        if not self._queue:
                            break
                        batch = self._queue[: self._batch_size]
                        self._queue = self._queue[self._batch_size :]
                    self._send_batch(batch)
            except Exception:
                logger.exception("ZROKY flush error")

    def _send_batch(self, events: list[dict[str, Any]]) -> int:
        """Send a batch of events, respecting the circuit breaker."""
        if not events:
            return 0

        # Circuit breaker check
        if self._circuit_open_until > time.time():
            logger.debug("ZROKY circuit open, requeueing %d events", len(events))
            with self._lock:
                self._queue = events + self._queue
            return 0

        try:
            if len(events) == 1:
                self._client.send_event(events[0])
            else:
                self._client.send_batch(events)

            self._consecutive_failures = 0
            return len(events)

        except Exception:
            self._consecutive_failures += 1
            if self._consecutive_failures >= _CIRCUIT_BREAKER_THRESHOLD:
                self._circuit_open_until = time.time() + _CIRCUIT_BREAKER_COOLDOWN
                logger.warning(
                    "ZROKY circuit breaker opened after %d failures, cooldown %ds",
                    _CIRCUIT_BREAKER_THRESHOLD,
                    _CIRCUIT_BREAKER_COOLDOWN,
                )

            if self._fail_open:
                logger.warning("ZROKY send failed, requeueing %d events", len(events))
                with self._lock:
                    self._queue = events + self._queue
                return 0
            raise

    @property
    def queue_size(self) -> int:
        """Current number of events in the queue."""
        with self._lock:
            return len(self._queue)

    @property
    def is_circuit_open(self) -> bool:
        """Whether the circuit breaker is currently open."""
        return self._circuit_open_until > time.time()

File: packages/sdk-node/src/monitor.ts
/**
 * ZROKY Monitor — high-level async event tracker with batching and fail-open.
 */

import { ZrokyClient, ZrokyClientConfig } from './client';

const DEFAULT_BATCH_SIZE = 50;
const DEFAULT_FLUSH_INTERVAL = 5_000;
const DEFAULT_MAX_QUEUE = 10_000;
const CIRCUIT_BREAKER_THRESHOLD = 5;
const CIRCUIT_BREAKER_COOLDOWN = 60_000;

export interface MonitorConfig extends ZrokyClientConfig {
  agentId: string;
  failOpen?: boolean;
  batchSize?: number;
  flushInterval?: number;
  maxQueueSize?: number;
}

export interface TrackOptions {
  model?: string;
  sessionId?: string;
  metadata?: Record<string, unknown>;
}

export type TrackResult =
  | { status: 'queued' }
  | { status: 'dropped'; reason: string };

interface QueuedEvent {
  agent_id: string;
  prompt: string;
  response: string;
  model: string;
  session_id: string;
  metadata: Record<string, unknown>;
  timestamp: number;
}

export class ZROKYMonitor {
  private readonly client: ZrokyClient;
  private readonly agentId: string;
  private readonly failOpen: boolean;
  private readonly batchSize: number;
  private readonly maxQueueSize: number;
  private queue: QueuedEvent[] = [];
  private closed = false;
  private flushTimer: ReturnType<typeof setInterval>;

  // Circuit breaker
  private consecutiveFailures = 0;
  private circuitOpenUntil = 0;

  constructor(config: MonitorConfig) {
    if (!config.agentId) {
      throw new Error('agentId is required');
    }

    this.agentId = config.agentId;
    this.failOpen = config.failOpen ?? true;
    this.batchSize = config.batchSize || DEFAULT_BATCH_SIZE;
    this.maxQueueSize = config.maxQueueSize || DEFAULT_MAX_QUEUE;
    this.client = new ZrokyClient(config);

    const interval = config.flushInterval || DEFAULT_FLUSH_INTERVAL;
    this.flushTimer = setInterval(() => this.flushAll(), interval);
    // Allow Node.js to exit even if timer is running
    if (this.flushTimer.unref) {
      this.flushTimer.unref();
    }
  }

  track(prompt: string, response: string, options: TrackOptions = {}): TrackResult {
    if (this.closed) {
      throw new Error('Monitor is closed');
    }

    if (this.queue.length >= this.maxQueueSize) {
      if (this.failOpen) {
        return { status: 'dropped', reason: 'queue_full' };
      }
      throw new Error(`ZROKY event queue full (${this.maxQueueSize})`);
    }

    this.queue.push({
      agent_id: this.agentId,
      prompt,
      response,
      model: options.model || '',
      session_id: options.sessionId || '',
      metadata: options.metadata || {},
      timestamp: Date.now() / 1000,
    });

    return { status: 'queued' };
  }

  async flush(): Promise<number> {
    if (this.queue.length === 0) return 0;

    const batch = this.queue.splice(0, this.batchSize);
    return this.sendBatch(batch);
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    clearInterval(this.flushTimer);

    while (this.queue.length > 0) {
      const batch = this.queue.splice(0, this.batchSize);
      await this.sendBatch(batch);
    }
  }

  get queueSize(): number {
    return this.queue.length;
  }

  get isCircuitOpen(): boolean {
    return this.circuitOpenUntil > Date.now();
  }

  // -- Internal ------------------------------------------------------------

  private async flushAll(): Promise<void> {
    try {
      while (this.queue.length > 0) {
        const batch = this.queue.splice(0, this.batchSize);
        await this.sendBatch(batch);
      }
    } catch {
      // fail-open: swallow errors in background flush
    }
  }

  private async sendBatch(events: QueuedEvent[]): Promise<number> {
    if (events.length === 0) return 0;

    // Circuit breaker check
    if (this.circuitOpenUntil > Date.now()) {
      this.queue.unshift(...events);
      return 0;
    }

    try {
      if (events.length === 1) {
        await this.client.sendEvent(events[0] as unknown as Record<string, unknown>);
      } else {
        await this.client.sendBatch(
          events as unknown as Record<string, unknown>[],
        );
      }
      this.consecutiveFailures = 0;
      return events.length;
    } catch {
      this.consecutiveFailures++;
      if (this.consecutiveFailures >= CIRCUIT_BREAKER_THRESHOLD) {
        this.circuitOpenUntil = Date.now() + CIRCUIT_BREAKER_COOLDOWN;
      }

      if (this.failOpen) {
        this.queue.unshift(...events);
        return 0;
      }
      throw new Error('ZROKY send failed');
    }
  }
}
